rackspace/autoscale/v1/policies/results.go

```diff
@@ -0,0 +1,138 @@
+package policies
+
+import (
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/pagination"
+)
+
+type policyResult struct {
+	gophercloud.Result
+}
+
+// Extract interprets any policyResult as a Policy, if possible.
+func (r policyResult) Extract() (*Policy, error) {
+	if r.Err != nil {
+		return nil, r.Err
+	}
+
+	var response struct {
+		Policy Policy `mapstructure:"policy"`
+	}
+
+	err := mapstructure.Decode(r.Body, &response)
+
+	return &response.Policy, err
+}
+
+// CreateResult represents the result of a create operation.
+type CreateResult struct {
+	policyResult
+}
+
+// Extract extracts a slice of Policies from a CreateResult.  Multiple policies
+// can be created in a single operation, so the result of a create is always a
+// list of policies.
+func (res CreateResult) Extract() ([]Policy, error) {
+	if res.Err != nil {
+		return nil, res.Err
+	}
+
+	return commonExtractPolicies(res.Body)
+}
+
+// GetResult temporarily contains the response from a Get call.
+type GetResult struct {
+	policyResult
+}
+
+// UpdateResult represents the result of an update operation.
+type UpdateResult struct {
+	gophercloud.ErrResult
+}
+
+// DeleteResult represents the result of a delete operation.
+type DeleteResult struct {
+	gophercloud.ErrResult
+}
+
+// ExecuteResult represents the result of an execute operation.
+type ExecuteResult struct {
+	gophercloud.ErrResult
+}
+
+// Policy represents a scaling policy.
+type Policy struct {
+	// UUID for the policy.
+	ID string `mapstructure:"id" json:"id"`
+
+	// Name of the policy.
+	Name string `mapstructure:"name" json:"name"`
+
+	// Type of scaling policy.
+	Type Type `mapstructure:"type" json:"type"`
+
+	// Cooldown period, in seconds.
+	Cooldown int `mapstructure:"cooldown" json:"cooldown"`
+
+	// Number of servers added or, if negative, removed.
+	Change interface{} `mapstructure:"change" json:"change"`
```

Comment:
Thinking about this more. Not sure this is doable with mapstructure. It's easy enough adding an `UnmarshalJSON` method to `Policy`, but I think that means doing one of the following:

- Keep the raw HTTP response in the `Body` field on the `gophercloud.Result` structs and do the decoding in the extract methods.

- Do the decoding in the request methods, and store a `Policy` in `gophercloud.Result.Body`. The extract methods don't do much then.

- Add a `Policy` to the `policyResult` struct and store the decoded policies there.

The first option seems most reasonable to me. Anyway, my hangup is that all of those break from the norm of decoding into a generic map on `gophercloud.Result.Body` and then using mapstructure. I can't find anywhere else in the code that does something like this.

Thoughts? Also, if it's easier, I can push something I have along the lines of the first option above and we can talk about it more concretely maybe.